### PR TITLE
Keaton allow alias within select

### PR DESCRIPTION
**Pull request by keatoooon**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 1]
```diff
@@ -6418,6 +6430,17 @@ public SqlNode expandGroupByOrHavingOrQualifyExpr(SqlNode expr,
     return newExpr;
   }
 
+  public SqlNode expandWithAlias(SqlNode expr,
```
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:6445]: Can you a docstring here since this seems to be the real implementation that's used?

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 2]
```diff
@@ -7204,6 +7227,146 @@ protected SqlNode expandDynamicStar(SqlIdentifier id, SqlIdentifier fqId) {
...
+    SqlNode expr = null;
+    int numOccurrencesOfAlias = 0;
+    int idxOfOccurrence = 0;
+
+    for (int i = 0; i < maxIdx; i++) {
+      SqlNode s = select.getSelectList().get(i);
+      final String alias = SqlValidatorUtil.getAlias(s, -1);
+      if (alias != null && nameMatcher.matches(alias, name)) {
+        expr = s;
+        idxOfOccurrence = i;
```
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:7281]: What happens if this matches more than once? Is it not a problem if idxOfOccurrence gets clobbered?

Comment (njriasan) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:7281]: An example would also be really helpful here.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 3]
```diff
@@ -7204,6 +7227,146 @@ protected SqlNode expandDynamicStar(SqlIdentifier id, SqlIdentifier fqId) {
...
+  /**
+   * Shuttle which walks over an expression replacing usage of alias with underlying expression.
+   */
+  static class ExtendedAliasExpander extends Expander {
+    SqlSelect aliasSourceSelect;
+
+    // When recursively checking for aliases, we often encounter situations where we need to check
+    // a subset of the select items for an alias. IE:
+    //    SELECT A, B, C from ...
+    // If we're trying to find aliases for "B" we should only check every select statement that
```
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: What happens is the alias is seen multiple times? If its ambiguous it should raise an exception, but I think mimicking your example below:

```SQL
SELECT C as x, x, A + x from
```

this should be okay.

Comment (njriasan) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: This is probably an edge case though so if we throw an exception even in this case that would be fine with me.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 4]
```diff
@@ -7204,6 +7227,146 @@ protected SqlNode expandDynamicStar(SqlIdentifier id, SqlIdentifier fqId) {
...
+      this.aliasSourceSelect = select;
+      this.maxIdx = select.getSelectList().size();
+    }
+
+
+    ExtendedAliasExpander(SqlValidatorImpl validator, SqlValidatorScope scope,
+        SqlSelect select, Integer maxIdx) {
+      // This constructor is used when attempting to expand a select item, since we need to check
+      // a subset of the select list for alias information
+      // Note that it is also called recursivley, in order to handle expanding chains of aliases,
```
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: recursivley [sic]

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 5]
```diff
@@ -7204,6 +7227,146 @@ protected SqlNode expandDynamicStar(SqlIdentifier id, SqlIdentifier fqId) {
...
+    //    SELECT A, x, C as x from ...
+    // but the following statement will work:
+    //    SELECT C as x, x, A from ...
+    //
+    // Note that we must also handle the following case (which is allowed in SF):
+    //    SELECT A as x, x as x from ...
+
+    // This argument instructs the expander to only check elements in the select list
+    // up to but not including this index.
+    Integer maxIdx;
```
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Minor nitpick. The comment is clear but `maxIdx` may imply inclusive to some people. Maybe `maxNumCols` or something indicating its a length would be more obvious.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 6]
```diff
@@ -7204,6 +7227,146 @@ protected SqlNode expandDynamicStar(SqlIdentifier id, SqlIdentifier fqId) {
...
+      super(validator, scope);
+      this.aliasSourceSelect = select;
+      this.maxIdx = maxIdx;
+      assert maxIdx <= select.getSelectList().size() && maxIdx >= 0;
+    }
+
+
+
+    @Override public @Nullable SqlNode visit(SqlIdentifier id) {
+      if (id.isSimple()) {
```
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Why do we require the id to be simple? This seems like it needs a comment.

Comment (njriasan) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Is it a limitation in the implementation or a fundamental reason?

Comment (njriasan) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: I assume this is because an alias is always a simple name and if we can't resolve an id that is in the same projection it must be an alias?

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 7]
```diff
@@ -7204,6 +7227,146 @@ protected SqlNode expandDynamicStar(SqlIdentifier id, SqlIdentifier fqId) {
...
+          int numOccurrencesOfAlias = occurenceNumExprAndExprIdx.left;
+
+
+          if (numOccurrencesOfAlias == 0) {
+            //If we can't find any aliases in the select list, just throw
+            // the original error.
+            throw e;
+          } else if (numOccurrencesOfAlias > 1) {
+            // More than one column has this alias.
+            throw validator.newValidationError(id,
```
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Adding a comment to make sure this is tested.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 8]
```diff
@@ -7364,18 +7527,30 @@ static class ExtendedExpander extends Expander {
           && (havingExpr
               ? validator.config().conformance().isHavingAlias()
               : !groupByExpr || validator.config().conformance().isGroupByAlias())) {
+
+        // Note: for Groupby/having, aliases in the select list take precedence over
+        // columns in the source table. This is the reverse of the precedence
+        // for where/on clauses.
+
+        // NOTE: this is not True for SF:
```
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: If this is unclear is it possible to throw an exception in the case in may be ambiguous?

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 9]
```diff
@@ -1168,18 +1168,28 @@ private void convertWhere(
...
     final RelFactories.FilterFactory filterFactory =
         RelFactories.DEFAULT_FILTER_FACTORY;
-    final RelNode filter =
-        filterFactory.createFilter(bb.root(), convertedWhere2, ImmutableSet.of());
+    final RelNode filter;
+    try {
+      filter =
+          filterFactory.createFilter(bb.root(), convertedWhere2, ImmutableSet.of());
+    } catch (Throwable e) {
+      System.out.println("TODO!");
```
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java]: What is this TODO meant to handle? Does this need to be a proper exception?

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 10]
```diff
@@ -1168,18 +1168,28 @@ private void convertWhere(
     SqlNode newWhere = pushDownNotForIn(bb.scope(), where);
     replaceSubQueries(bb, newWhere, RelOptUtil.Logic.UNKNOWN_AS_FALSE);
     final RexNode convertedWhere = bb.convertExpression(newWhere);
-    final RexNode convertedWhere2 =
+    RexNode convertedWhere2 =
```
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java]: Revert this change. It can still be final.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 11]
```diff
@@ -2721,6 +2721,46 @@ private void checkHavingAliasSameAsColumn(boolean upperAlias) {
     sql(query).ok(expected);
   }
 
+  @Test public void testSelectQueryAliasInWhereClause() {
```
Comment (njriasan) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:2724]: Let's comment out any tests that don't pass yet.

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java [hunk 12]
```diff
@@ -5186,6 +5186,324 @@ void checkCorrelatedMapSubQuery(boolean expand) {
     assertThat(parameters.get(1).getType().getSqlTypeName(), is(SqlTypeName.VARCHAR));
   }
 
+  //TODO: resolving in a followup issue: https://bodo.atlassian.net/browse/BE-4092
+  @Test public void testAliasCommonExpressionPushDownWhere() {
```
Comment (njriasan) [inline core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java]: Same comment as before.

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java [hunk 13]
```diff
@@ -8950,25 +8950,78 @@ public void _testGroupExpressionEquivalenceParams() {
...
+  // ... as opposed to what was there before:
+  //    final SqlNode expandedWhere = expand(where, whereScope);
+  //
+  // At first glance fix to this is unclear to me, since we will have ambiguous situations where
+  // we will need to check table columns to resolve aliases, and we cannot rely on strictly
+  // identifier names.
+  // (see https://docs.google.com/document/d/1bxBLuH-4gB9E5zAgTbv4a7soHHfa1jErlXAkg9a8G0M/edit)
+  //
+  // I'm also not certain why this is being tested, or why it would benefit us to not throw
+  // an error in this situation. Honestly, I think we definitely SHOULD throw an error during
```
Comment (njriasan) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:8964]: Can we update this test to throw an error?

**Review by njriasan — COMMENTED** [review]
Overall this looks good, I just think we need more error handling tests.

In particular I would like to see:

1. A test for the ambiguous alias situation.
2. Any test that currently doesn't pass so have a corresponding replacement test that checks that there is a reasonable error message.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 3 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: The behavior (both in SF and locally) is that a failure, with the message that x is an ambiguous column name. I'll add a test for it.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 6 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: correct. I'll add a comment to this effect.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 7 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: tested in SqlValidatorTest (see testFailsAmbiguous and testFailsAmbiguous2)

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java (hunk 9 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java]: Sorry, this is left over debugging code, removing.

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml [hunk 14]
```diff
@@ -4179,7 +4465,7 @@ when matched and target.sal > 10 then
 when matched and target.sal < 10 then
   update set sal = target.sal + source.sal + 10
 when matched and target.sal = 10 then
-  update set sal = -10
+  DELETE
```
Comment (keatoooon) [inline core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:4468]: There's a few changes to SQL strings, which occurred when I copied the "actual" generated XML file into this one.

**Review by njriasan — APPROVED** [review]
LGTM!
